tasty-query/shared/src/main/scala/tastyquery/Subtyping.scala

```diff
@@ -0,0 +1,306 @@
+      isSubtype(tp1.first, tp2) && isSubtype(tp1.second, tp2)
+
+    case tp1: AnnotatedType =>
+      isSubtype(tp1.typ, tp2)
+
+    case _ =>
+      level3(tp1, tp2)
+  end level2
+
+  private def level3(tp1: Type, tp2: Type)(using Context): Boolean = tp2 match
+    case tp2: TypeRef =>
+      tp2.symbol match
+        case cls2: ClassSymbol =>
+          if cls2.typeParams.isEmpty then
+            if isBottom(tp1) then true
+            else if tp1.isLambdaSub then false // should be tp1.hasHigherKind, but the scalalib does not like that
+            else if cls2 == defn.AnyClass then true
+            else level3WithBaseType(tp1, tp2, cls2)
+          else
+            // TODO Try eta-expansion if tp1.isLambdaSub && !tp1.isAnyKind
+            level4(tp1, tp2)
+        case sym2: TypeSymbolWithBounds =>
+          isSubtype(tp1, tp2.prefix.memberTypeBoundsLow(sym2))
+            || level4(tp1, tp2)
+      end match
+
+    case tp2: AppliedType =>
+      compareAppliedType2(tp1, tp2)
+
+    case tp2: OrType =>
+      isSubtype(tp1, tp2.first) || isSubtype(tp1, tp2.second)
+
+    case _ =>
+      level4(tp1, tp2)
+  end level3
+
+  private def level3WithBaseType(tp1: Type, tp2: Type, cls2: ClassSymbol)(using Context): Boolean =
+    val base = tp1.baseType(cls2)
+    if base != NoType && (base ne tp1) then isSubtype(base, tp2)
+    else level4(tp1, tp2)
+  end level3WithBaseType
+
+  private def compareAppliedType2(tp1: Type, tp2: AppliedType)(using Context): Boolean =
+    val tycon2 = tp2.tycon
+    val tparams = tycon2.typeParams
+    if tparams.isEmpty then
+      throw InvalidProgramStructureException(s"found type constructor $tycon2 without type params in AppliedType")
+
+    def isMatchingApply(tp1: Type): Boolean = tp1.widen match
+      case tp1: AppliedType =>
+        tp1.tycon match
+          case tycon1: TypeRef =>
+            tp2.tycon match
+              case tycon2: TypeRef =>
+                val tycon1Sym = tycon1.symbol
+                val tycon2Sym = tycon2.symbol
+                if tycon1Sym == tycon2Sym && isSubprefix(tycon1.prefix, tycon2.prefix) then
+                  isSubArgs(tp1.args, tp2.args, tp1, tparams)
+                else false
+              case _ =>
+                false
+        end match
+
+      case _ =>
+        false
+    end isMatchingApply
+
+    tp2.tycon match
+      case tycon2: TypeRef =>
+        if isMatchingApply(tp1) then true
+        else
+          tycon2.symbol match
+            case cls2: ClassSymbol =>
+              // TODO Handle generic tuple <: TupleN
+              level3WithBaseType(tp1, tp2, cls2)
+            case sym2: TypeSymbolWithBounds =>
+              // TODO Handle polymorphic type aliases (compareLower in TypeComparer)
+              level4(tp1, tp2)
+        end if
+
+      case _ =>
+        false
+  end compareAppliedType2
+
+  private def isSubArgs(args1: List[Type], args2: List[Type], tp1: AppliedType, tparams2: List[TypeParamInfo])(
```

Comment:
tp1 is unused here (in the original implementation it looks like it's only used for capture calculus stuff)